src/cli/telemetry/telemetry.ts

```diff
@@ -0,0 +1,293 @@
+    os_name: details.platform,
+    os_version: details.release,
+    cpu_model: `${details.cpuModel}`,
+    build: `${result.buildId}`,
+    typescript: versions.typescript,
+    rollup: versions.rollup,
+  };
+
+  await sendMetric('stencil_cli_command', data);
+  logger.debug(`${logger.blue('Telemetry')}: ${logger.gray(JSON.stringify(data))}`);
+}
+
+export async function telemetryAction(options: telemetryActionArgs) {
+  const { args, sys, logger, config, action } = options;
+  const tracking = await shouldTrack(!!config.flags.ci);
+
+  const coreCompiler = await loadCoreCompiler(sys);
+  const details = sys.details;
+  const versions = coreCompiler.versions;
+  const flags = parseFlags(args, sys);
+  let duration = undefined;
+  let error: any;
+
+  if (action) {
+    const start = new Date();
+
+    try {
+      await action();
+    } catch (e) {
+      error = e;
+    }
+
+    const end = new Date();
+    duration = end.getTime() - start.getTime();
+  }
+
+  // We'll get componentCount details inside the taskBuild, so let's not send two messages.
+  if (!tracking || (flags.task == 'build' && !flags.args.includes('--watch'))) {
+    return;
+  }
+
+  const session_id = await getTelemetryToken();
+  const packages = await getInstalledPackages();
+  const targets = await getActiveTargets(config);
+
+  const data: TrackableData = {
+    session_id,
+    yarn: isUsingYarn(),
+    duration,
+    targets,
+    packages: packages || [],
+    arguments: flags.args,
+    task: flags.task,
+    stencil: coreCompiler.version,
+    system: `${sys.name} ${sys.version}`,
+    os_name: details.platform,
+    os_version: details.release,
+    cpu_model: `${details.cpuModel}`,
+    build: coreCompiler.buildId,
+    typescript: versions.typescript,
+    rollup: versions.rollup,
+  };
+
+  await sendMetric('stencil_cli_command', data);
+  logger.debug(`${logger.blue('Telemetry')}: ${logger.gray(JSON.stringify(data))}`);
+
+  if (error) {
+    throw error;
+  }
+}
+
+function isUsingYarn() {
+  return process.env.npm_execpath.includes('yarn');
+}
+
+async function getActiveTargets(config: Config): Promise<string[][] | OutputTargetOptions[][]> {
+  return (config.outputTargets as any[]).map(target => {
+    let outputTargetSettings: OutputTargetOptions = {};
+
+    if (target.type === 'www') {
+      outputTargetSettings.serviceWorker = isObject(target?.serviceWorker);
+    }
+
+    return [target.type, outputTargetSettings];
+  });
+}
+
+async function getInstalledPackages() {
+  try {
+    // Read package.json and package-lock.json
+    const appRootDir = process.cwd();
+
+    const packageJson: PackageJsonData = await tryFn(
+      readJson,
+      getCompilerSystem().resolvePath(appRootDir + '/package.json'),
+    );
+    const packageLockJson: any = await tryFn(
+      readJson,
+      getCompilerSystem().resolvePath(appRootDir + '/package-lock.json'),
+    );
+
+    const packages = Object.entries({
+      ...packageJson.devDependencies,
```

Comment:
It really _shouldn't_ be, but what happens when `tryFn` returns `null` in the event of a read failure? Does this cause an `Error` trying to access a property on `undefined`?

I think that the type declaration on [this line](https://github.com/ionic-team/stencil/pull/2964/files#diff-2a84eb0343da5e51e060e0697c4c6fb67c2cc98b591ca94599ac984573787c3cR184) is tricking the compiler into thinking this can never be `null`, but I'm not sure that's the case.

I would recommend returning an empty array & logging the error in the event we cannot read their `package.json` or `package-lock.json`

Reply:
I'm pretty sure I added some adequate error handling here. Let me know if you can think of anything else.